### ISSUE TITLE
GRPO training collapses after ~270 steps — reward drops to zero

### ISSUE THREAD

**Issue by dukesun99** [BeastyZ/ConvSearch-R1#6]
Hi, thank you so much for this excellent work! The idea of applying R1-style reasoning to conversational search is really compelling, and I appreciate you open-sourcing the full training pipeline — it's been very helpful for understanding the approach.

I'm trying to reproduce the GRPO training on QReCC with Qwen2.5-3B-Instruct, following the exact hyperparameters from the paper and the example scripts. Training consistently collapses around step 270-300 — the mean reward/score drops from ~1.3 to 0, and the model starts outputting random multilingual gibberish.

My setup:

- 2x GPUs (vs 8x in the paper), but the effective batch size is identical since train_batch_size=128 is global and FSDP + gradient accumulation compensates
- All hyperparameters match the paper exactly: lr=1e-6, warmup=100, kl_loss_coef=0.001, clip_ratio=0.2, n=8, temperature=0.7, train_batch_size=128
- SFT model trained with the provided scripts (batch_size=64, max_length=3072, 2 epochs, lr=1e-5)
- Collapse occurs at ~1.2 epochs (29596 / 128 ≈ 231 steps/epoch), reproducible across multiple runs

Training curves:

<img width="881" height="325" alt="Image" src="https://github.com/user-attachments/assets/9adebdf8-4f11-4ab5-a5a2-db12354d8b00" />

- Orange / Pink (kl_loss_coef=0.001, paper setting): both collapse at ~step 270-300, reward drops to 0
- Blue (kl_loss_coef=0.02): stable training past step 450+, reward stays ~1.3-1.5

Increasing the KL loss coefficient by 20x resolves the collapse, but this differs from the paper's reported value of 0.001.

Questions:

1. The cur_rewrite_reward / cur_cal_reward curriculum reward is fully implemented in rewrite_r1.py, but the call is commented out in ray_trainer.py (line 878). Was the curriculum reward used in the actual experiments? 
2. The language_reward function is defined in rewrite_r1.py but not called from compute_score. Was it used during training?
3. Were there any other stabilization techniques not reflected in the released scripts?

Any guidance on reproducing stable training with the paper's kl_loss_coef=0.001 would be greatly appreciated. Happy to share full wandb logs if that helps. Thanks again for the great work!

**Comment by BeastyZ** [comment]
Welcome to follow our work!

Before answering your questions, may I ask: are you using BM25 to provide the reward? If so, please switch to ANCE, as BM25 is too sensitive to text and will cause unstable training. In our work, all RL is trained with rewards provided by ANCE, and then evaluated on both ANCE and BM25 separately.

Below are the answers to your three questions:
1. cur_rewrite_reward / cur_cal_reward were experimental attempts and were **not** used in the final reported results. The runnable code shown in the current repository is exactly the code that produces the main results in the paper.
2. As mentioned above, BM25 is sensitive to text and causes training instability. We therefore tried introducing language consistency to mitigate this issue, but it did not work. You can thus ignore the `language_reward` function.
3. We recommend using only ANCE to provide the reward for RL during training.

Please check the Wandb logs here: [Qwen-Qrecc](https://api.wandb.ai/links/1015602635-fudan-university/sjbg152k) & [Qwen-Topiocqa](https://api.wandb.ai/links/1015602635-fudan-university/etm7jaz3)

**Comment by BeastyZ** [comment]
Close the current issue. If you have any questions, please reopen the issue.

**Comment by dukesun99** [comment]
Thank you! Yes, it was the issue of the retriever that I did not notice. May I also ask if you have tried any more recent retriever models? How do they compare to ANCE?

**Comment by BeastyZ** [comment]
> Thank you! Yes, it was the issue of the retriever that I did not notice. May I also ask if you have tried any more recent retriever models? How do they compare to ANCE?

To facilitate comparison with previous work, most of our experiments were conducted based on ANCE. During the peer review stage, the reviewers requested us to test more recent retrievers. We therefore carried out additional experiments on RepLLaMA (which, while not the latest retriever available today, is newer than ANCE), as shown in Table 8 of our [paper](https://arxiv.org/pdf/2505.15776). The overall conclusion is that retrievers with stronger inherent performance generally yield better downstream results.

**Comment by dukesun99** [comment]
> > Thank you! Yes, it was the issue of the retriever that I did not notice. May I also ask if you have tried any more recent retriever models? How do they compare to ANCE?
> 
> To facilitate comparison with previous work, most of our experiments were conducted based on ANCE. During the peer review stage, the reviewers requested us to test more recent retrievers. We therefore carried out additional experiments on RepLLaMA (which, while not the latest retriever available today, is newer than ANCE), as shown in Table 8 of our [paper](https://arxiv.org/pdf/2505.15776). The overall conclusion is that retrievers with stronger inherent performance generally yield better downstream results.

Thank you for the quick reply. May I confirm is RepLLaMA only tested in downstream tasks or also used for providing rewards for the results in Table 8?

**Comment by BeastyZ** [comment]
> > > Thank you! Yes, it was the issue of the retriever that I did not notice. May I also ask if you have tried any more recent retriever models? How do they compare to ANCE?
> > 
> > 
> > To facilitate comparison with previous work, most of our experiments were conducted based on ANCE. During the peer review stage, the reviewers requested us to test more recent retrievers. We therefore carried out additional experiments on RepLLaMA (which, while not the latest retriever available today, is newer than ANCE), as shown in Table 8 of our [paper](https://arxiv.org/pdf/2505.15776). The overall conclusion is that retrievers with stronger inherent performance generally yield better downstream results.
> 
> Thank you for the quick reply. May I confirm is RepLLaMA only tested in downstream tasks or also used for providing rewards for the results in Table 8?

RepLLaMA is also used for providing rewards for the results in Table 8.
